kraken/src/models/file/operations.rs

```diff
@@ -0,0 +1,70 @@
+use rorm::and;
+use rorm::db::transaction::Transaction;
+use rorm::insert;
+use rorm::prelude::ForeignModel;
+use rorm::prelude::ForeignModelByField;
+use rorm::prelude::*;
+use rorm::query;
+use rorm::Patch;
+use uuid::Uuid;
+
+use crate::models::MediaFile;
+use crate::models::User;
+use crate::models::Workspace;
+
+impl MediaFile {
+    /// Inserts a new [`MediaFile`]
+    ///
+    /// ## Beware:
+    /// - This function checks for duplicates using `name` and `sha256`.
+    ///   If it finds one, it won't insert the new file but return the old file's uuid.
+    ///   In that case, the new file has to be deleted on disk.
+    /// - The insert should only be applied after every filesystem operation finished successfully.
+    ///   But the duplicate should be checked as soon as the entire file is available.
+    ///   So the insert has to be run in a transaction whose commit is delayed until everything else finished.
+    pub async fn get_or_insert(
+        tx: &mut Transaction,
```

Comment:
Just take an `impl Executor<'_>` and ensure the transaction internally so for externals usage, this API looks like a single operation.

Reply:
No, because like the doc comment explains, the transaction has to be delayed by the caller
i.e. passing a `Database` instead of a `Transaction` will ALWAYS be a logic error.
It makes no sense for the signature to allow it in the first place.